### PR TITLE
[Akka.Streams] make default `LogSource`s actually usable

### LINKED ISSUE #7126: Akka.Streams: make it easier to pass in a friendly `LogSource` name
**Is your feature request related to a problem? Please describe.**

Right now the built in `.Log` method in Akka.Streams prints out a lot of junk under the `LogSource` of one of the built-in materialized actors. It should be easier to pass in a human-friendly log source name as an `ActorAttribute` when organizing the stream.

**Describe the solution you'd like**

```csharp
var task = source.DivertTo(evenSink, i => i % 2 == 0)
.AddAttributes(ActorAttributes.CreateLogSource(...))
.DivertTo(threeSink, i => i % 3 == 0)
.DivertTo(fiveSink, i => i % 5 == 0)
.RunWith(Sink.ForEach<int>(i => Console.WriteLine($"{i} is unhandled")), actorSystem);
```

**Describe alternatives you've considered**

I'm going to make a separate issue about the `.Log` method itself in Akka.Streams - maybe we can make that method customizable instead.

### PR DESCRIPTION
**Pull request by Aaronontheweb**
## Changes

Fixes #7126 

So TL;DR;, logging from within an Akka.Streams graph stage is basically unusable and has been since the inception of Akka.Streams.

Behold, what my logs look like from my Akka.Streams-powered MQTT client library:

```text
[DEBUG][04/24/2024 00:50:51.923Z][Thread 0046][akka://test/user/turbomqtt-clients/tcp] Creating new TCP transport for [CreateTcpTransport { Options = MqttClientTcpOptions { AddressFamily = Unspecified, MaxFrameSize = 131072, Host = localhost, Port = 21883, ReconnectInterval = 00:00:05, MaxReconnectAttempts = 10 }, ProtocolVersion = V3_1_1 }]
[DEBUG][04/24/2024 00:50:51.924Z][Thread 0046][akka://test/user/turbomqtt-clients/tcp/$a] Created new TCP transport for client connecting to [localhost:21883]
[INFO][04/24/2024 00:50:51.928Z][Thread 0013][akka://test/user/turbomqtt-clients/tcp/$a] Attempting to connect to [localhost:21883]
[DEBUG][04/24/2024 00:50:51.929Z][Thread 0037][akka://test/user/turbomqtt-clients/tcp/$a] Attempting to connect to [localhost:21883] - resolved to [::1, 127.0.0.1]
[INFO][04/24/2024 00:50:51.930Z][Thread 0037][akka://test/user/turbomqtt-clients/tcp/$a] Successfully connected to [localhost:21883]
[DEBUG][04/24/2024 00:50:51.930Z][Thread 0013][LogSource (akka://test)] Encoded 1 messages using 37 bytes
[DEBUG][04/24/2024 00:50:51.930Z][Thread 0046][TcpMqtt311End2EndSpecs (akka://test)] Decoded 1 packets from transport.
[INFO][04/24/2024 00:50:51.930Z][Thread 0046][TcpMqtt311End2EndSpecs (akka://test)] Received packet of type Connect
[DEBUG][04/24/2024 00:50:51.930Z][Thread 0046][TcpMqtt311End2EndSpecs (akka://test)] Sending packet of type ConnAck using V3_1_1
[DEBUG][04/24/2024 00:50:51.932Z][Thread 0046][TcpMqtt311End2EndSpecs (akka://test)] Successfully wrote packet of type ConnAck [4 bytes] to transport.
[DEBUG][04/24/2024 00:50:51.932Z][Thread 0013][LogSource (akka://test)] Decoded [1] packets totaling [4] bytes
[DEBUG][04/24/2024 00:50:51.933Z][Thread 0013][LogSource (akka://test)] Received packet of type [ConnAck] from client.
[DEBUG][04/24/2024 00:50:51.933Z][Thread 0046][LogSource (akka://test)] Encoded 1 messages using 12 bytes
[DEBUG][04/24/2024 00:50:51.933Z][Thread 0013][TcpMqtt311End2EndSpecs (akka://test)] Decoded 1 packets from transport.
[INFO][04/24/2024 00:50:51.933Z][Thread 0013][TcpMqtt311End2EndSpecs (akka://test)] Received packet of type Subscribe
[DEBUG][04/24/2024 00:50:51.933Z][Thread 0013][TcpMqtt311End2EndSpecs (akka://test)] Sending packet of type SubAck using V3_1_1
[DEBUG][04/24/2024 00:50:51.933Z][Thread 0013][TcpMqtt311End2EndSpecs (akka://test)] Successfully wrote packet of type SubAck [5 bytes] to transport.
[DEBUG][04/24/2024 00:50:51.933Z][Thread 0032][LogSource (akka://test)] Decoded [1] packets totaling [5] bytes
[DEBUG][04/24/2024 00:50:51.933Z][Thread 0032][LogSource (akka://test)] Received packet of type [SubAck] from client.
[DEBUG][04/24/2024 00:50:51.934Z][Thread 0013][LogSource (akka://test)] Encoded 1 messages using 24 bytes
[WARNING][04/24/2024 00:50:51.934Z][Thread 0039][TcpMqtt311End2EndSpecs (akka://test)] Client test-client is being disconnected from server.
[DEBUG][04/24/2024 00:50:51.934Z][Thread 0039][TcpMqtt311End2EndSpecs (akka://test)] Sending packet of type Disconnect using V3_1_1
[ERROR][04/24/2024 00:50:51.934Z][Thread 0039][TcpMqtt311End2EndSpecs (akka://test)] Failed to write packet of type Disconnect to transport.
[ERROR][04/24/2024 00:50:51.934Z][Thread 0037][akka://test/user/turbomqtt-clients/tcp/$a] Failed to read from socket.
```

There are about 10 different stream stages in there, each of which get created twice (rebooting a failed connection) - I can't tell _any_ of them apart, and I know I have a bug in that code somewhere.

Thus, I've made this PR and now these logs will look more like this going forward:

```text
[INFO][04/24/2024 02:25:06.096Z][Thread 0016][FlowShape`2([LogStage.in] [LogStage.out])(akka://AkkaStreamsLogSourceSpec-1/user/StreamSupervisor-0)] Element: 1
[INFO][04/24/2024 02:25:06.097Z][Thread 0016][FlowShape`2([LogStage.in] [LogStage.out])(akka://AkkaStreamsLogSourceSpec-1/user/StreamSupervisor-0)] Element: 2
[INFO][04/24/2024 02:25:06.097Z][Thread 0016][FlowShape`2([LogStage.in] [LogStage.out])(akka://AkkaStreamsLogSourceSpec-1/user/StreamSupervisor-0)] Element: 3
[INFO][04/24/2024 02:25:06.098Z][Thread 0016][FlowShape`2([LogStage.in] [LogStage.out])(akka://AkkaStreamsLogSourceSpec-1/user/StreamSupervisor-0)] Element: 4
[INFO][04/24/2024 02:25:06.098Z][Thread 0016][FlowShape`2([LogStage.in] [LogStage.out])(akka://AkkaStreamsLogSourceSpec-1/user/StreamSupervisor-0)] Element: 5
[DEBUG][04/24/2024 02:25:06.119Z][Thread 0012][CoordinatedShutdown (akka://AkkaStreamsLogSourceSpec-1)] Performing phase [before-service-unbind] with [0] tasks.
[DEBUG][04/24/2024 02:25:06.119Z][Thread 0015][CoordinatedShutdown (akka://AkkaStreamsLogSourceSpec-1)] Performing phase [service-unbind] with [0] tasks.
[DEBUG][04/24/2024 02:25:06.120Z][Thread 0007][CoordinatedShutdown (akka://AkkaStreamsLogSourceSpec-1)] Performing phase [service-requests-done] with [0] tasks.
[DEBUG][04/24/2024 02:25:06.120Z][Thread 0017][CoordinatedShutdown (akka://AkkaStreamsLogSourceSpec-1)] Performing phase [service-stop] with [0] tasks.
[DEBUG][04/24/2024 02:25:06.120Z][Thread 0017][CoordinatedShutdown (akka://AkkaStreamsLogSourceSpec-1)] Performing phase [before-cluster-shutdown] with [0] tasks.
```

The default `LogSource` for any Akka.Streams stage, including built in ones is now `{shape.ToString()}({streamSupervisorActorPath})`.

The `shape.ToString()` method does the following by default:

```csharp
public sealed override string ToString() => $"{GetType().Name}([{string.Join(", ", Inlets)}] [{string.Join(", ", Outlets)}])";
```

Which will print out the type of stage and a rough description of its inputs and outputs, in accordance with the names given to those Inlet and Outlet types. That will add a small amount of overhead to each stage we create (a `.ToString()` call that iterates over some types that return hard-coded strings back) but in return you can now immediately understand:

1. Where in the actor hierarchy this stream is
2. What type of stage it is
3. What its inputs and outputs are

Moreover, if you override the `GraphStage.LogSource` with a friendlier name we will still append the `StreamSupervisor.Path` to it during logger construction, so you can still keep track of where this stream is within the actor hierarchy.

TL;DR; this is a substantial usability improvement for Akka.Streams.

## Checklist

For significant changes, please ensure that the following have been completed (delete if not relevant):

* [x] This change follows the [Akka.NET API Compatibility Guidelines](https://getakka.net/community/contributing/api-changes-compatibility.html).
* [x] This change follows the [Akka.NET Wire Compatibility Guidelines](https://getakka.net/community/contributing/wire-compatibility.html).
* [x] I have [reviewed my own pull request](https://getakka.net/community/contributing/index.html#review-your-own-pull-requests).
* [x] Design discussion issue #7126 
* [x] Changes in public API reviewed, if any.

### CODE REVIEW

File: src/core/Akka.API.Tests/verify/CoreAPISpec.ApproveCore.DotNet.verified.txt [hunk 1]
```diff
@@ -3440,6 +3440,7 @@ namespace Akka.Event
         public System.Type Type { get; }
         public static Akka.Event.LogSource Create(object o) { }
         public static Akka.Event.LogSource Create(object o, Akka.Actor.ActorSystem system) { }
+        public static Akka.Event.LogSource Create(string source, System.Type t) { }
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/verify/CoreAPISpec.ApproveCore.DotNet.verified.txt:3443]: Public API - this basically calls the CTOR of `LogSource`.

File: src/core/Akka.Streams.Tests/AkkaStreamsLogSourceSpec.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+        await probe.ExpectMsgAsync(1);
+        await probe.ExpectMsgAsync(2);
+        await probe.ExpectMsgAsync(3);
+        await probe.ExpectMsgAsync(4);
+        await probe.ExpectMsgAsync(5);
+        
+        // check just the first log message
+        var logMessage = logProbe.ExpectMsg<Debug>();
+        logMessage.LogSource.Should().Contain("StreamSupervisor");
```
Comment (Aaronontheweb) [inline src/core/Akka.Streams.Tests/AkkaStreamsLogSourceSpec.cs:94]: Laying down some regression tests here to make sure someone doesn't cleverly optimize-away useful data out of the `LogSource` name for these stages.

File: src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs [hunk 3]
```diff
@@ -384,7 +384,13 @@ public override TMat Materialize<TMat>(IGraph<ClosedShape, TMat> runnable, Func<
         /// </summary>
         /// <param name="logSource">The source that produces the log events.</param>
         /// <returns>The newly created logging adapter.</returns>
-        public override ILoggingAdapter MakeLogger(object logSource) => Logging.GetLogger(System, logSource);
+        public override ILoggingAdapter MakeLogger(object logSource)
+        {
+            if (logSource is not LogSource s) return Logging.GetLogger(System, logSource);
```
Comment (Aaronontheweb) [inline src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs]: Use the old code if someone's not using a `LogSource` - I'm open to changing this so everything does it the new way.

File: src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs [hunk 4]
```diff
@@ -402,7 +408,7 @@ public override void Shutdown()
                 Supervisor.Tell(PoisonPill.Instance);
         }
 
-        private ILoggingAdapter GetLogger() => _system.Log;
+        private ILoggingAdapter GetLogger() => MakeLogger(_supervisor);
```
Comment (Aaronontheweb) [inline src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs:419]: Always create a new logger, don't filch and just use the `ActorSystem`'s one.

File: src/core/Akka.Streams/Implementation/Fusing/Ops.cs [hunk 5]
```diff
@@ -2981,17 +2983,7 @@ public override void PreStart()
...
-                        var materializer = ActorMaterializerHelper.Downcast(Materializer);
-                        _log = new BusLogging(materializer.System.EventStream, _stage._name, GetType(), materializer.System.Settings.LogFormatter);
-                    }
-                    catch (Exception ex)
-                    {
-                        throw new Exception(
-                            "Log stage can only provide LoggingAdapter when used with ActorMaterializer! Provide a LoggingAdapter explicitly or use the actor based flow materializer.",
-                            ex);
-                    }
+                    _log = Log;
```
Comment (Aaronontheweb) [inline src/core/Akka.Streams/Implementation/Fusing/Ops.cs:2986]: Simplify the built-in `.Log` stage to just use the built-in `Log` rather than doing all of this weird stuff.

File: src/core/Akka.Streams/Stage/GraphStage.cs [hunk 6]
```diff
@@ -840,7 +840,7 @@ protected GraphStageLogic(int inCount, int outCount)
         /// <param name="shape">TBD</param>
         protected GraphStageLogic(Shape shape) : this(shape.Inlets.Count(), shape.Outlets.Count())
         {
-            LogSource = Akka.Event.LogSource.Create(shape);
+            LogSource = Akka.Event.LogSource.Create(shape.ToString());
```
Comment (Aaronontheweb) [inline src/core/Akka.Streams/Stage/GraphStage.cs:843]: Get the full details on the `shape` so we can see input and output ports too.

File: src/core/Akka/Event/Logging.cs [hunk 7]
```diff
@@ -64,18 +64,25 @@ public static LogSource Create(object o, ActorSystem system)
                     return new LogSource(FromActorRef(actorRef, system), SourceType(actorRef));
                 case string str:
                     return new LogSource(FromString(str, system), SourceType(str));
-                case System.Type t:
+                case Type t:
                     return new LogSource(FromType(t, system), t);
+                case LogSource logSource:
+                    return logSource; // if someone's already created a LogSource, just use it
```
Comment (Aaronontheweb) [inline src/core/Akka/Event/Logging.cs:70]: If someone calls `LogSource.Create` with a ... `LogSource`... as input, just use the `LogSource`. This was "eating" user-defined `LogSource` names before.

**Review by Aaronontheweb — COMMENTED** [review]
Detailed my changes

File: src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs (hunk 3 above)
Comment (Aaronontheweb) [reply src/core/Akka.Streams/Implementation/ActorMaterializerImpl.cs]: Done, went ahead and made this change.

**Review by Arkatufus — APPROVED** [review]
LGTM

File: src/core/Akka.Streams/Implementation/Fusing/Ops.cs (hunk 5 above)
Comment (Danthar) [reply src/core/Akka.Streams/Implementation/Fusing/Ops.cs:2986]: where does `Log` come from ? The BusLogging previously made sure that the messages are communicated through the EventStream system to the user Logging Actor. Thus not causing a blocking call here.
Does this do the same ?

Comment (Aaronontheweb) [reply src/core/Akka.Streams/Implementation/Fusing/Ops.cs:2986]: @Danthar it comes from the `GraphStage.Log` property, which is the abstract base class for all stream shapes. It's lazily instantiated and gets created by the `ActorMaterializer` only when it's called.

Comment (Aaronontheweb) [reply src/core/Akka.Streams/Implementation/Fusing/Ops.cs:2986]: It's still using `BusLogging` et al - it's just doing a more manual version of `Context.GetLogger()` because there's no guarantee that the the `StageRef` for this `GraphStage` is available (for reasons I don't quite understand, but that came up during testing.)

Comment (Danthar) [reply src/core/Akka.Streams/Implementation/Fusing/Ops.cs:2986]: Ah yep found it.
That call eventually ends up at:
       `     return new BusLogging(context.System.EventStream, logSource.Source, logSource.Type, logMessageFormatter ?? context.System.Settings.LogFormatter);`
            
So to answer my own question. Yes it does seems to do the same.
